framework/include/utils/MathFVUtils.h

```diff
@@ -241,6 +243,87 @@ linearInterpolation(const T & value1,
   return coeffs.first * value1 + coeffs.second * value2;
 }
 
+/**
+ * This is a helper function for the harmonic interpolation to ensure we throw a meaningful
+ * compile-time error if the developer is trying to use harmonic interpolation on an
+ * unsupported type
+ */
+template <class T>
+constexpr std::false_type always_false{};
+
+/**
+ * Computes the harmonic mean (1/(gc/value1+(1-gc)/value2)) of Reals, RealVectorValues and
+ * RealTensorValues with accounting for the possibility that one or both of them are AD.
+ * For tensors, we use a component-wise mean instead of the matrix-inverse based option.
+ */
+template <typename T1, typename T2>
+typename libMesh::CompareTypes<T1, T2>::supertype
+harmonicInterpolation(const T1 & value1,
+                      const T2 & value2,
+                      const FaceInfo & fi,
+                      const bool one_is_elem)
+{
+  // We check if the base values of the given template types match, if not we throw a compile-time
+  // error
+  static_assert(std::is_same<typename MetaPhysicL::RawType<T1>::value_type,
+                             typename MetaPhysicL::RawType<T2>::value_type>::value,
+                "The input values for harmonic interpolation need to have the same raw-value!");
+
+  // Fetch the interpolation coefficients, we use the exact same coefficients as for a simple
+  // weighted average
+  const auto coeffs = interpCoeffs(InterpMethod::Average, fi, one_is_elem);
+
+  // We check if the types are fit to compute the harmonic mean of. This is done compile-time
+  // using constexpr
+  if constexpr (std::is_same<typename MetaPhysicL::RawType<T1>::value_type, Real>::value)
```

Comment:
Just to make this more readable, let's comment each section, i.e.:

- `// Real case`
- `// RealVectorValue case`
- `// RealTensorValue case`
- `// YouScrewedUpHard case`